### ISSUE TITLE
Memory leak on Windows

### ISSUE THREAD

**Issue by os12** [grpc/grpc#8236]
I've just noticed a slow yet steady leak on Windows, when serving RPCs with 1..128 kB payloads (as strings in protos). This happens when the Windows host is a server.

The leak happens here:

```
+   11484 ( 3ce78 - 2b9f4)     7e allocs	BackTrace5847B5BF
+      19 (    7e -    65)	BackTrace5847B5BF	allocations

	ntdll!LdrOpenImageFileOptionsKey+53
	ntdll!RtlAllocateHeap+1182
	cohesity_windows_agent_service!heap_alloc_dbg_internal+1FA (d:\th\minkernel\crts\ucrt\src\appcrt\heap\debug_heap.cpp, 359)
	cohesity_windows_agent_service!heap_alloc_dbg+4D (d:\th\minkernel\crts\ucrt\src\appcrt\heap\debug_heap.cpp, 450)
	cohesity_windows_agent_service!_malloc_dbg+2F (d:\th\minkernel\crts\ucrt\src\appcrt\heap\debug_heap.cpp, 492)
	cohesity_windows_agent_service!malloc+1E (d:\th\minkernel\crts\ucrt\src\appcrt\heap\malloc.cpp, 22)
	cohesity_windows_agent_service!gpr_malloc+4B (grpc-1.0.0\src\core\lib\support\alloc.c, 58)
	cohesity_windows_agent_service!win_write+177 (grpc-1.0.0\src\core\lib\iomgr\tcp_windows.c, 298)
	cohesity_windows_agent_service!grpc_endpoint_write+62 (grpc-1.0.0\src\core\lib\iomgr\endpoint.c, 44)
	cohesity_windows_agent_service!grpc_chttp2_perform_writes+120 (grpc-1.0.0\src\core\ext\transport\chttp2\transport\writing.c, 193)
	cohesity_windows_agent_service!writing_action+66 (grpc-1.0.0\src\core\ext\transport\chttp2\transport\chttp2_transport.c, 994)
	cohesity_windows_agent_service!???+0 : 7FF62401DD81
	cohesity_windows_agent_service!grpc_pollset_work+14E (grpc-1.0.0\src\core\lib\iomgr\pollset_windows.c, 153)
	cohesity_windows_agent_service!???+0 : 7FF623FFE6AE
	cohesity_windows_agent_service!grpc::CompletionQueue::AsyncNextInternal+73 (grpc-1.0.0\src\cpp\common\completion_queue.cc, 56)
	cohesity_windows_agent_service!grpc::CompletionQueue::Next+8E (c:\dev\toolchain\x86_64-windows\grpc-1.0.0\include\grpc++\impl\codegen\completion_queue.h, 147)
	cohesity_windows_agent_service!<lambda_3280da030d3cdf4fa699535b10d53763>::operator()+82 (agent_rpc_async_service.cpp, 215)
	cohesity_windows_agent_service!std::_Invoker_functor::_Call<<lambda_3280da030d3cdf4fa699535b10d53763> & __ptr64>+3E (c:\program files (x86)\microsoft visual studio 14.0\vc\include\type_traits, 1377)
	cohesity_windows_agent_service!std::invoke<<lambda_3280da030d3cdf4fa699535b10d53763> & __ptr64>+3E (c:\program files (x86)\microsoft visual studio 14.0\vc\include\type_traits, 1445)
	cohesity_windows_agent_service!std::_Invoke_ret<void,<lambda_3280da030d3cdf4fa699535b10d53763> & __ptr64>+42 (c:\program files (x86)\microsoft visual studio 14.0\vc\include\type_traits, 1462)
	cohesity_windows_agent_service!std::_Func_impl<<lambda_3280da030d3cdf4fa699535b10d53763>,std::allocator<int>,void>::_Do_call+45 (c:\program files (x86)\microsoft visual studio 14.0\vc\include\functional, 214)
	cohesity_windows_agent_service!std::_Func_class<void>::operator()+69 (c:\program files (x86)\microsoft visual studio 14.0\vc\include\functional, 280)
	cohesity_windows_agent_service!cohesity::magneto::agents::windows::`anonymous namespace'::ThreadpoolCallback+48 (thread_pool.cpp, 32)
	ntdll!RtlReleaseSRWLockExclusive+35CE
	ntdll!RtlReleaseSRWLockExclusive+17E4
	KERNEL32!BaseThreadInitThunk+14
	ntdll!RtlUserThreadStart+21
```

One workaround is to disable the synchronous send attempt as it does not cleanup something properly: https://github.com/os12/grpc/commit/0dcd609be2ea3f9416048e425726503f107815a5

Yet, this should be fixed correctly and fully. Also, are there other places that do this `WSASend()` trick? Could someone have a look please?

Thanks!

**Comment by nicolasnoble** [comment]
Ah, thanks for the report.

**Comment by nicolasnoble** [comment]
Looks like github no longer closes issues when pull requests are merged...

That should be fixed in #8237.

**Comment by os12** [comment]
Thanks, @nicolasnoble - the patch works.
